stdlib/public/core/KeyPathIterable.swift

```diff
@@ -0,0 +1,115 @@
+//===-- KeyPathIterable.swift ---------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the KeyPathIterable protocol.
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// KeyPathIterable
+//===----------------------------------------------------------------------===//
+
+/// An implementation detail of `KeyPathIterable`; do not use this protocol
+/// directly.
+public protocol _KeyPathIterableBase {
+  var _allKeyPathsTypeErased: [AnyKeyPath] { get }
+  var _recursivelyAllKeyPathsTypeErased: [AnyKeyPath] { get }
+}
+
+/// A type whose values provides custom key paths to properties or elements.
+public protocol KeyPathIterable : _KeyPathIterableBase {
+  /// A type that can represent a collection of all key paths of this type.
+  associatedtype AllKeyPaths : Collection
+    where AllKeyPaths.Element == PartialKeyPath<Self>
+
+  /// A collection of all custom key paths of this value.
+  var allKeyPaths: AllKeyPaths { get }
+}
+
+public extension KeyPathIterable {
+  /// An array of all custom key paths of this value and any custom key paths
+  /// nested within each of what this value's key paths refers to.
+  var recursivelyAllKeyPaths: [PartialKeyPath<Self>] {
+    var result: [PartialKeyPath<Self>] = []
+    for kp in allKeyPaths {
+      result.append(kp)
+      if let nested = self[keyPath: kp] as? _KeyPathIterableBase {
+        for nkp in nested._recursivelyAllKeyPathsTypeErased {
+          result.append(kp.appending(path: nkp)!)
+        }
+      }
+    }
+    return result
+  }
+}
+
+extension KeyPathIterable {
+  public var _allKeyPathsTypeErased: [AnyKeyPath] {
+    return allKeyPaths.map { $0 as AnyKeyPath }
+  }
+  public var _recursivelyAllKeyPathsTypeErased: [AnyKeyPath] {
+    return recursivelyAllKeyPaths.map { $0 as AnyKeyPath }
+  }
+}
+
+public extension KeyPathIterable {
+  /// Returns an array of all custom key paths of this value, to the specified
+  /// type.
+  @inlinable
```

Comment:
There's no reason to serialize it right now. Let's remove `@inlinable` (and other occurrences) and reevaluate this later. When the synthesis can directly emit a flattened array literal, we should then serialize it.

Reply:
That said, the required computed property (which you synthesize) should be `@inlinable`.

Reply:
Added `@inlinable` to synthesized `allKeyPaths` in https://github.com/apple/swift/pull/21557/commits/a0635fa7f8314c7af117066780877c01d018dbfb. Verified that `@inlinable` exists in SILGen.